### ISSUE TITLE
pnpm stuck on 7.1.7 under corepack

### ISSUE THREAD

**Issue by hyperknot** [nodejs/corepack#225]
Copying pnpm issue here: https://github.com/pnpm/pnpm/issues/5875

### Code to reproduce the issue: 
```
corepack prepare pnpm@7.22.0 --activate; pnpm -v
```

### Expected behavior:
It should output I'm on 7.22.0

### Actual behavior:
```
Preparing pnpm@7.22.0 for immediate activation...
7.1.7
```

### Additional information:

 - `node -v` prints: v16.17.1
 - Windows, macOS, or Linux?: macOS, nvm

```
which pnpm
~/.nvm/versions/node/v16.17.1/bin/pnpm
```

**Comment by merceyz** [comment]
Have you run`corepack enable`? Your reproduction is missing it;
```console
$ docker run --rm node:16.17.1 bash -c "corepack prepare pnpm@7.22.0 --activate; pnpm -v"
Preparing pnpm@7.22.0 for immediate activation...
bash: pnpm: command not found
```

**Comment by hyperknot** [comment]
Yes, I've done corepack enable the first time I installed it, otherwise I wouldn't have been able to use pnpm.

But now, I've run it again, just for testing and after that I'm able to do the upgrade.

So the bug is basically this: To update pnpm, I need to repeatedly run `corepack enable`, before each update.

**Comment by arcanis** [comment]
Something happened that caused the binary to be overwritten. It's not possible to say what with the info you have, but Corepack doesn't remove them by itself, so it points to a third-party issue.

Some Node.js version managers in particular are known to forget the Corepack status when switching between Node versions.

**Comment by styfle** [comment]
Also try `echo $PATH` or `echo %PATH%` which will show you the precedence of which directories are searched on your machine. As mentioned above, likely nvm path is prepended in front of the corepack path which is causing it to prefer nvm.

**Comment by hyperknot** [comment]
The path is good, it's the same as it was when the bug was present.

```
which pnpm
~/.nvm/versions/node/v16.17.1/bin/pnpm
```

I definitely haven't done anything with global packages. What I have done though is I removed some of my cache folders after migrating to pnpm. What does `corepack enable` do exactly? Could I have ruined it by deleting my cache folders?

**Comment by aduh95** [comment]
> What does `corepack enable` do exactly?

It adds the binaries to your path, next to the `node` one:

https://github.com/nodejs/corepack/blob/5633a158986bae193a737b82b08d0139e078694b/README.md#L128-L138

If you're using nvm, I'm not sure how this works as I believe it exposes different `node` binary depending on the context.

> Could I have broken it by deleting my cache folders?

What cache folders?

**Comment by hyperknot** [comment]
Just the npm, yarn, etc. cache folders I found under home dir and Library.

**Comment by mrhammadasif** [comment]
I would like to add that the `corepack` looks for the `packageManager` property in package.json, Just make sure to update that as well to the latest version.

**Comment by iapetos163** [comment]
I was facing the same issue and it turns our my problem was due to the `package.json` configuration. I ran this command

```
$ corepack prepare pnpm@8.6.1 --activate
Preparing pnpm@8.6.1 for immediate activation...
```

Despite the command and its output, corepack actually installs `pnpm@7.1.8` because that's what's specified in `package.json`.

This is very confusing behavior. I would consider it more intuitive for corepack to always install an explicitly-given version, and maybe output a warning if the version in `package.json` does not match.
